File: packages/providers/src/cursor.ts
import type { ProviderUsage } from "./types.js";
import { clampPercent, makeWindow, safeFetch } from "./utils.js";

interface CursorPlanUsage {
  totalSpend?: number;
  includedSpend?: number;
  limit?: number;
  totalPercentUsed?: number;
  autoPercentUsed?: number;
  apiPercentUsed?: number;
}

interface CursorPeriodUsage {
  planUsage?: CursorPlanUsage;
  billingCycleEnd?: string;
  membershipType?: string;
}

interface CursorUsageSummary {
  billingCycleEnd?: string;
  membershipType?: string;
  individualUsage?: {
    plan?: {
      used?: number;
      limit?: number;
      remaining?: number;
      totalPercentUsed?: number;
      apiPercentUsed?: number;
      autoPercentUsed?: number;
    };
  };
}

export async function fetchCursorUsage(
  sessionToken?: string
): Promise<ProviderUsage> {
  const base: ProviderUsage = {
    id: "cursor",
    name: "Cursor",
    status: "disconnected",
    windows: [],
    lastSyncedAt: new Date().toISOString(),
    accent: "#7C3AED",
  };

  if (!sessionToken) {
    return { ...base, error: "No session token found" };
  }

  const headers = {
    Cookie: `WorkosCursorSessionToken=${sessionToken}`,
    "Content-Type": "application/json",
  };

  const period = await safeFetch<CursorPeriodUsage>(
    "https://api2.cursor.sh/aiserver.v1.DashboardService/GetCurrentPeriodUsage",
    {
      method: "POST",
      headers,
      body: "{}",
    }
  );

  const summary = await safeFetch<CursorUsageSummary>(
    "https://www.cursor.com/api/usage-summary",
    { headers: { Cookie: headers.Cookie } }
  );

  const windows: ProviderUsage["windows"] = [];

  if (period.ok && period.data.planUsage) {
    const plan = period.data.planUsage;
    const limit = (plan.limit ?? 0) / 100;
    const used = (plan.totalSpend ?? plan.includedSpend ?? 0) / 100;

    if (limit > 0) {
      windows.push(
        makeWindow(
          "billing_cycle",
          "Monthly included",
          used,
          limit,
          period.data.billingCycleEnd
        )
      );
    }

    if (plan.totalPercentUsed != null) {
      windows.push({
        kind: "monthly",
        label: "Total usage",
        used: clampPercent(plan.totalPercentUsed),
        limit: 100,
        remaining: Math.max(0, 100 - clampPercent(plan.totalPercentUsed)),
        percentUsed: clampPercent(plan.totalPercentUsed),
        resetsAt: period.data.billingCycleEnd,
      });
    }

    if (plan.apiPercentUsed != null) {
      windows.push({
        kind: "monthly",
        label: "API models",
        used: clampPercent(plan.apiPercentUsed),
        limit: 100,
        remaining: Math.max(0, 100 - clampPercent(plan.apiPercentUsed)),
        percentUsed: clampPercent(plan.apiPercentUsed),
      });
    }

    if (plan.autoPercentUsed != null) {
      windows.push({
        kind: "monthly",
        label: "Cursor models",
        used: clampPercent(plan.autoPercentUsed),
        limit: 100,
        remaining: Math.max(0, 100 - clampPercent(plan.autoPercentUsed)),
        percentUsed: clampPercent(plan.autoPercentUsed),
      });
    }
  }

  if (summary.ok && summary.data.individualUsage?.plan) {
    const plan = summary.data.individualUsage.plan;
    if (plan.limit && plan.used != null && windows.length === 0) {
      windows.push(
        makeWindow(
          "billing_cycle",
          "Plan requests",
          plan.used,
          plan.limit,
          summary.data.billingCycleEnd
        )
      );
    }
  }

  if (windows.length === 0) {
    return {
      ...base,
      status: "error",
      error: period.ok
        ? "Connected but no usage windows returned"
        : period.error,
    };
  }

  return {
    ...base,
    status: "connected",
    plan: period.ok
      ? period.data.membershipType
      : summary.ok
        ? summary.data.membershipType
        : undefined,
    windows,
  };
}

File: apps/desktop/src-tauri/src/providers/cursor.rs
use serde_json::Value;

use super::{window_dollars, window_percent};
use crate::types::ProviderUsage;

pub async fn fetch(token: Option<String>) -> ProviderUsage {
    let base = ProviderUsage {
        id: "cursor".into(),
        name: "Cursor".into(),
        plan: None,
        status: "disconnected".into(),
        windows: vec![],
        last_synced_at: chrono::Utc::now().to_rfc3339(),
        error: None,
        accent: "#7C3AED".into(),
    };

    let token = match token {
        Some(t) if !t.is_empty() => t,
        _ => return ProviderUsage { error: Some("Not connected".into()), ..base },
    };

    let client = reqwest::Client::new();
    let cookie = format!("WorkosCursorSessionToken={}", token);

    let summary = client
        .get("https://www.cursor.com/api/usage-summary")
        .header("Cookie", cookie.clone())
        .send()
        .await;

    let mut windows = vec![];

    if let Ok(resp) = summary {
        if let Ok(json) = resp.json::<Value>().await {
            if let Some(plan) = json.pointer("/individualUsage/plan") {
                let used = plan.get("used").and_then(|v| v.as_f64());
                let limit = plan.get("limit").and_then(|v| v.as_f64());
                let resets = json
                    .get("billingCycleEnd")
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string());

                if let (Some(used), Some(limit)) = (used, limit) {
                    windows.push(window_dollars("billing_cycle", "Monthly included", used, limit, resets));
                }

                if let Some(pct) = plan.get("apiPercentUsed").and_then(|v| v.as_f64()) {
                    windows.push(window_percent("monthly", "API models", pct, None));
                }
                if let Some(pct) = plan.get("autoPercentUsed").and_then(|v| v.as_f64()) {
                    windows.push(window_percent("monthly", "Cursor models", pct, None));
                }
            }
        }
    }

    if windows.is_empty() {
        return ProviderUsage {
            status: "error".into(),
            error: Some("Could not read Cursor usage. Try reconnecting.".into()),
            ..base
        };
    }

    ProviderUsage {
        status: "connected".into(),
        plan: Some("Pro".into()),
        windows,
        ..base
    }
}
